runtime/cache/cache.go

```diff
@@ -0,0 +1,156 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/micro/go-micro/v3/runtime"
+)
+
+// NewCache wraps a runtime with a cache
+func NewCache(r runtime.Runtime) runtime.Runtime {
+	return &cache{
+		lastUpdated: make(map[string]time.Time),
+		services:    make(map[string][]*runtime.Service),
+		mux:         new(sync.RWMutex),
+		Runtime:     r,
+	}
+}
+
+type cache struct {
+	// lastUpdated contains the last time services were read from the underlying runtime for a given
+	// namespace. When a service is Created/Updated/Deleted in a given namespace, the value is deleted
+	// from the map
+	lastUpdated map[string]time.Time
+	// services is a cache of the services in a namespace, it's reset when the lastUpdated is reset
+	services map[string][]*runtime.Service
+	// mux is a mutex to protect the lastUpdated and services
+	mux *sync.RWMutex
+
+	runtime.Runtime
+}
+
+// Create a service
+func (c *cache) Create(srv *runtime.Service, opts ...runtime.CreateOption) error {
+	// parse the options
+	var options runtime.CreateOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if options.Namespace == "" {
+		options.Namespace = "micro"
+	}
+
+	// call the underlying runtime
+	if err := c.Runtime.Create(srv, opts...); err != nil {
+		return err
+	}
+
+	// the service was written so reset the cache for the namespace
+	c.mux.Lock()
+	delete(c.lastUpdated, options.Namespace)
+	delete(c.services, options.Namespace)
+	c.mux.Unlock()
+
+	return nil
+}
+
+// Read returns the service
+func (c *cache) Read(opts ...runtime.ReadOption) ([]*runtime.Service, error) {
+	// parse the options
+	var options runtime.ReadOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if options.Namespace == "" {
+		options.Namespace = "micro"
+	}
+
+	// if a query was used we can't use the cache
+	if len(options.Service) > 0 || len(options.Type) > 0 || len(options.Version) > 0 {
```

Comment:
if there are this many cases where we can't use the cache is it even worth having it?

Reply:
Yes, although these options exist, the vast majority of callers won't pass this option.

Reply:
Also, these options are usually used in tandem with each other